threeML/plugins/OGIP/eventlist.py

```diff
@@ -388,190 +408,225 @@ def _fit_background(self):
         # The total cnts (over channels) is binned to 1 sec intervals
-        if self._poly_order == -1:
-            totalbkgevents = self._arrival_times[background_mask]
-            binwidth = 1.
+        if self._user_poly_order == -1:
+            totalbkgevents = self._arrival_times[poly_mask]
+            bin_width = .1
             cnts, bins = np.histogram(totalbkgevents,
                                       bins=np.arange(self._start_time,
                                                      self._stop_time,
-                                                     binwidth))
+                                                     bin_width))
 
-            cnts = cnts / binwidth
+            cnts = cnts / bin_width
             # Find the mean time of the bins
-            meantime = []
+            mean_time = []
             for i in xrange(len(bins) - 1):
                 m = np.mean((bins[i], bins[i + 1]))
-                meantime.append(m)
-            meantime = np.array(meantime)
+                mean_time.append(m)
+            mean_time = np.array(mean_time)
 
             # Remove bins with zero counts
-            allnonzeromask = []
+            all_non_zero_mask = []
+            
             for bkgsel in self._poly_time_selections:
-                allnonzeromask.append(np.logical_and(meantime >= bkgsel[0],
-                                                     meantime <= bkgsel[1]))
+                all_non_zero_mask.append(np.logical_and(mean_time >= bkgsel[0],
+                                                        mean_time <= bkgsel[1]))
+
+            non_zero_mask = all_non_zero_mask[0]
+            if len(all_non_zero_mask) > 1:
+                for mask in all_non_zero_mask[1:]:
+                    non_zero_mask = np.logical_or(mask, non_zero_mask)
 
-            nonzeromask = allnonzeromask[0]
-            if len(allnonzeromask) > 1:
-                for mask in allnonzeromask[1:]:
-                    nonzeromask = np.logical_or(mask, nonzeromask)
+            self._optimal_polynomial_grade = self._fit_global_and_determine_optimum_grade(cnts[non_zero_mask],
+                                                                                          mean_time[non_zero_mask])
+
+            print("Auto-determined polynomial order: %d" % self._optimal_polynomial_grade)
+            print
 
-            self.optimalPolGrade = self._fit_global_and_determine_optimum_grade(cnts[nonzeromask],
-                                                                                meantime[nonzeromask])
 
         else:
 
-            self.optimalPolGrade = self._poly_order
+            self._optimal_polynomial_grade = self._user_poly_order
 
-            print "Best fit polynomial order:"
-            print "\t: %d" % self.optimalPolGrade
+        # Attempting a parallel execution
 
+        channels = range(self._first_channel, self._n_channels + self._first_channel)
+        num_cpus = multiprocessing.cpu_count()
```

Comment:
The way of implementing parallel computation in 3ML is through the use of the Client from parallel/parallel_client.py . Look at the computation of contours in classicMLE/joint_likelihood.py for an example.

The reason to use that is that it allows a lot more flexibility through ipyparallel, for example allowing processes to be spawn across computer nodes. It also allows the user to cherry-pick what he wants to parallelize.

For example, if you run 100 jobs on a computer farm, you need to know in advance how many CPU each job will use, so the job should not decide by itself to use all the cpus of the node in which is running.

Please port this part to the Client interface, thanks!